sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LeafNode, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{MULTI_SCALAR_SUBQUERY, SCALAR_SUBQUERY}
+
+/**
+ * This rule tries to merge multiple non-correlated [[ScalarSubquery]]s into a
+ * [[MultiScalarSubquery]] to compute multiple scalar values once.
+ *
+ * The process is the following:
+ * - While traversing through the plan each [[ScalarSubquery]] plan is tried to merge into the cache
+ *   of already seen subquery plans. If merge is possible then cache is updated with the merged
+ *   subquery plan, if not then the new subquery plan is added to the cache.
+ * - The original [[ScalarSubquery]] expression is replaced to a reference pointing to its cached
+ *   version in this form: `GetStructField(MultiScalarSubquery(SubqueryReference(...)))`.
+ * - A second traversal checks if a [[SubqueryReference]] is pointing to a subquery plan that
+ *   returns multiple values and either replaces only [[SubqueryReference]] to the cached plan or
+ *   restores the whole expression to its original [[ScalarSubquery]] form.
+ * - [[ReuseSubquery]] rule makes sure that merged subqueries are computed once.
+ *
+ * Eg. the following query:
+ *
+ * SELECT
+ *   (SELECT avg(a) FROM t GROUP BY b),
+ *   (SELECT sum(b) FROM t GROUP BY b)
+ *
+ * is optimized from:
+ *
+ * Project [scalar-subquery#231 [] AS scalarsubquery()#241,
+ *   scalar-subquery#232 [] AS scalarsubquery()#242L]
+ * :  :- Aggregate [b#234], [avg(a#233) AS avg(a)#236]
+ * :  :  +- Relation default.t[a#233,b#234] parquet
+ * :  +- Aggregate [b#240], [sum(b#240) AS sum(b)#238L]
+ * :     +- Project [b#240]
+ * :        +- Relation default.t[a#239,b#240] parquet
```

Comment:
I think that is smart rewrite way but let's check a more complex example: E.g. can we rewrite 
```
SELECT *
FROM r
JOIN r2 ON r2.x = r.x
WHERE r.y = (SELECT sum(b) FROM t) AND r2.y = (SELECT avg(b) FROM t)
```
? Maybe
```
SELECT *
FROM (
  SELECT (
    SELECT STRUCT(sum(b) AS sum_b, avg(b) AS avg_b) FROM t) AS st, x, y 
    FROM r
  ) AS r
)
JOIN r2 ON r2.x = r.x
WHERE r.y = r.st.sum_b AND r2.y = r.st.avg_b
```
? Does this work with outer joins? And isn't this more complex than the reuse way in this PR?

I was also thinking about "whole plan subquery merge" (similar to my "whole plan reuse" suggestion: https://github.com/apache/spark/pull/28885) where subqueries at "different level" could be merged (and reused) as a possible improvement to this PR in the future.

BTW, the `ReuseExchangeAndSubquery` rule you mentioned is suggested in my "whole plan reuse" PR, which got stuck a bit due to lack of reviews. Do you also have a similar rule in production or you just saw my PR? If you have some time, any feedback is appreciated there as well. :)

I didn't check how correlated subqueries could benefit from rewriting the query (this PR focuses on uncorrelated ones), but I think at this point in the optimizer those have been transformed to joins.
Can you please elaborate on the "reading from tables v.s. reading from an array column" part?

Reply:
>> And isn't this more complex than the reuse way in this PR?

IIUC, I think this PR's complexity is in extra dependencies. The proposed rule augments two subqueries, makes them look identical, and *hopes* (a) column-pruning doesn't prune too aggressively and (b) physical de-dup could dedup them. In case (a) changes later and the two aggregate trees are not deduped in the physical plan, there could potentially be regressions -- each aggregation then becomes more expensive.

Therefore, I think there're two directions that you can go:
- (1) Add a top plan node that contains the main plan tree as well as CommonSubplan trees. For each merged subquery, put it into a CommonSubplan. In a main plan's node, it can reference a CommonSubplan and converts its output relation into a scalar value. This approach only works for uncorrelated cases.
- (2) Only extract common subqueries within the same logical node, and put merged scalar subqueries into a Project node below the logical node. It may be less ambitious than what you proposed, but it at least does not have extra dependencies. This approach works for both correlated and uncorrelated cases.

>> Can you please elaborate on the "reading from tables v.s. reading from an array column" part?

This is an example:

SELECT y
FROM LATERAL VIEW explode(ARRAY(ARRAY(1), ARRAY(1, 2), ARRAY(1, 2, 3))) AS y
WHERE 
      ( SELECT COUNT(*) FROM LATERAL VIEW explode(y) AS element WHERE element !=2 ) > 1
       AND 
      ( SELECT SUM(element) FROM LATERAL VIEW explode(y) AS element WHERE element !=1 ) > 3 

I noticed that such subqueries do not work for now. But they align with the language spec and has well defined semantics. Once we support them, we want your proposed rule to be able to speedup them as well.

>> ? Does this work with outer joins? 

I don't think outer joins are relevant to this problem:
- if we go with (1), it doesn't matter;
- if we go with (2), pushing projection expressions through join/outer join is separate topic/rule. A subquery expression doesn't have much difference from an ordinary expression in this regard.

Reply:
> The proposed rule augments two subqueries, makes them look identical, and hopes (a) column-pruning doesn't prune too aggressively and (b) physical de-dup could dedup them. In case (a) changes later and the two aggregate trees are not deduped in the physical plan, there could potentially be regressions -- each aggregation then becomes more expensive.

In this PR the new `MergeScalarSubqueries` rule runs in a separate batch after column pruning, close to the end of optimization. This is by design to make sure no subsequent rule changes the structure of different instances of a merged subquery plan at different places in the logical plan differently. So the physical planing creates the same physical plan for these instances and there shouldn't be any dedup issues.

I think probably the downside of my current PR is that the physical planning of merged subqueries happen multiple times (as many times as they they appear in the logical plan) and physical dedup comes only after that. This could be improved if we had subquery references in logical plan as well (something like `ReuseSubqueryExec`). But I think that's what your (1) is about. Move the merged subqueries to a special top logical plan node and add subquery references at places where they are actually used.

> SELECT y
FROM LATERAL VIEW explode(ARRAY(ARRAY(1), ARRAY(1, 2), ARRAY(1, 2, 3))) AS y
WHERE
( SELECT COUNT(*) FROM LATERAL VIEW explode(y) AS element ) > 1
AND
( SELECT SUM(element) FROM LATERAL VIEW explode(y) AS element ) > 3
I noticed that such subqueries do not work for now. But they align with the language spec and has well defined semantics. Once we support them, we want your proposed rule to be able to speedup them as well.

Ah ok, but what should be the optimized plan of that query? This looks like we have 2 correlated subqueries and (2) makes perfect sense to merge them. But I don't think we need lateral views, just take the following query:
```
SELECT
  (SELECT avg(a) FROM t WHERE t.a = outer.a),
  (SELECT sum(b) FROM t WHERE t.a = outer.a)
FROM t AS outer
```
which is
```
Project [avg(a)#236 AS scalarsubquery(a)#243, sum(b)#238L AS scalarsubquery(a)#244L]
+- Join LeftOuter, (a#241 = a#233)
   :- Project [a#233, avg(a)#236]
   :  +- Join LeftOuter, (a#239 = a#233)
   :     :- Project [a#233]
   :     :  +- Relation default.t[a#233,b#234] parquet
   :     +- Aggregate [a#239], [avg(a#239) AS avg(a)#236, a#239]
   :        +- Project [a#239]
   :           +- Filter isnotnull(a#239)
   :              +- Relation default.t[a#239,b#240] parquet
   +- Aggregate [a#241], [sum(b#242) AS sum(b)#238L, a#241]
      +- Filter isnotnull(a#241)
         +- Relation default.t[a#241,b#242] parquet
```
now, and this PR doesn't help at all, but it could be optimized using your (2).

I wonder the following steps (tickets/PRs) would make sense:

1. Finish this PR and support only on non-correlated mergeable subqueries. Mainly focus on merging plans and keep the physical reuse dependency for simplicity. This supports subquery merging within a plan regardless they are in the same logical node.
2. Add a performance improvement to 1. so as to physical plan a merged subquery only once. This is your (1) basically. Move the merged subqueries to a top node and introduce subquery references in logical plan.
3. Add support for correlated subqueries using your (2). As you mentioned this will only support subqueries within the same logical node.

Probably we should implement separate rules for 1. + 2. and 3. but the plan merging logic can be common.

Reply:
>> In this PR the new MergeScalarSubqueries rule runs in a separate batch 
>> after column pruning, close to the end of optimization. 
>> This is by design to make sure no subsequent rule changes the structure 

I don't see how we can guarantee that. I think the hidden, inter-rule dependency can add complexities for future development and maintenance. 

For instance, someone could implement a new Strategy that internally calls ColumnPruning after exploring one logical plan alternatives. By the time such a Strategy is implemented, the authors wouldn't be aware of the fact that ColumnPruning should *not* be called after MergeScalarSubqueries. When they find that's an issue, they would then have to either (a) add some hacks in the Aggregate to mark that MergeScalarSubqueries has been applied and hence ColumnPruning should not go through it, or (b) re-implement the rule per my proposal (1).
  
I'm wondering whether we can pursue (2) for now, if it meets your need. It's less ambitious but may address most of your issue? 

>> Add a performance improvement to 1. so as to physical plan a merged 
>> subquery only once. This is your (1) basically.

The performance was not my initial concern, but rather, we'd better make MergeScalarSubqueries self-contained and does not depend on *an assumption that could be changed*.

>>  But I don't think we need lateral views

Sub-querying over arrays are important use cases, for which we don't want to de-correlate. In this case, a subquery is more like an ordinary expression and should be evaluated within the operator node.

Reply:
> For instance, someone could implement a new Strategy that internally calls ColumnPruning after exploring one logical plan alternative. By the time such a Strategy is implemented, the authors wouldn't be aware of the fact that ColumnPruning should not be called after MergeScalarSubqueries.

I see, thanks. I've never seen such transformations in `SparkStrategy`s. 

But, if we followed (2) with non-correlated subqueries like this example:
```
SELECT t1.*
FROM t as t1
JOIN t as t2 ON t2.a = t1.a
WHERE
  t1.b = (SELECT sum(a) FROM t)
  AND t2.b = (SELECT count(a) FROM t)
```
and if I get your (2) right the rewritten query is:
```
SELECT t12.a, t12.b
FROM (
  SELECT
    t1.*,
    (SELECT STRUCT(sum(a) AS sum_a, count(a) AS count_a) FROM t) AS st,
    t1.b AS t1_b,
    t2.b AS t2_b
  FROM t as t1
  JOIN t as t2 ON t2.a = t1.a
) t12
WHERE
  t1_b = st.sum_a
  AND t2_b = st.count_a
```
so we basically add an extra project node under `Filter`. The analyzed plan is:
```
Project [a#237, b#238]
+- Filter ((cast(t1_b#235 as bigint) = st#234.sum_a) AND (cast(t2_b#236 as bigint) = st#234.count_a))
   +- SubqueryAlias t12
      +- Project [a#237, b#238, scalar-subquery#233 [] AS st#234, b#238 AS t1_b#235, b#240 AS t2_b#236]
         :  +- Aggregate [struct(sum_a, sum(a#244), count_a, count(a#244)) AS struct(sum(a) AS sum_a, count(a) AS count_a)#243]
         :     +- SubqueryAlias spark_catalog.default.t
         :        +- Relation default.t[a#244,b#245] parquet
         +- Join Inner, (a#239 = a#237)
            :- SubqueryAlias t1
            :  +- SubqueryAlias spark_catalog.default.t
            :     +- Relation default.t[a#237,b#238] parquet
            +- SubqueryAlias t2
               +- SubqueryAlias spark_catalog.default.t
                  +- Relation default.t[a#239,b#240] parquet
```
The optimzer (`PushDownPredicates`) would duplicate and push down the subquery under both sides of the join:
```
Project [a#237, b#238]
+- Join Inner, (a#239 = a#237)
   :- Filter ((isnotnull(b#238) AND (cast(b#238 as bigint) = scalar-subquery#233 [].sum_a)) AND isnotnull(a#237))
   :  :  +- Aggregate [struct(sum_a, sum(a#244), count_a, count(a#244)) AS struct(sum(a) AS sum_a, count(a) AS count_a)#243]
   :  :     +- Project [a#244]
   :  :        +- Relation default.t[a#244,b#245] parquet
   :  +- Relation default.t[a#237,b#238] parquet
   +- Project [a#239]
      +- Filter ((isnotnull(b#240) AND (cast(b#240 as bigint) = scalar-subquery#233 [].count_a)) AND isnotnull(a#239))
         :  +- Aggregate [struct(sum_a, sum(a#244), count_a, count(a#244)) AS struct(sum(a) AS sum_a, count(a) AS count_a)#243]
         :     +- Project [a#244]
         :        +- Relation default.t[a#244,b#245] parquet
         +- Relation default.t[a#239,b#240] parquet
```
Doesn't that mean that (2) also assumes that:
- no subsequent transformation changes the 2 instances differently and 
- `ReuseSubquery` does the dedup?

Reply:
>> Doesn't that mean that (2) also assumes that
>> no subsequent transformation changes the 2 instances differently and
>> ReuseSubquery does the dedup?

The difference is that transformation (2) is a self-contained transformation that results in a better plan, while this PR's MergeScalarSubqueries is not. If for some reason ReuseSubquery does not trigger, applying (2) does not make the plan worse than not applying (2). However, iiuc, in the same situation, MergeScalarSubqueries would make the plan worse because each aggregate pipeline runs additional computations.

Reply:
> If for some reason ReuseSubquery does not trigger, applying (2) does not make the plan worse than not applying (2).

In the above example the final optimized plan of (2) is the very same as with this PR. There are 2 aggregates in both subqueries so without dedup both (2) and this PR could cause regressions.
I agree that (2) is self-contained and this PR is not, but IMO it looks like there are inter-rule dependencies currently in Spark (like `PushDownPredicates` relies on `ReuseSubquery`) that overall doesn't make (2) safer than this PR.

I think this means that your (1) suggestion is probably the right approach and we need to move common non-correlated subqueries to a top node and reference to them in logical plan.

I also think that (2) is a good improvement for correlated subqueries, but I would pursue (1) in this PR first and maybe (2) in a separate one. Does this sound acceptable?

@cloud-fan, @maropu do you have any thoughts on this topic?

Reply:
>  I would pursue (1) in this PR first and maybe (2) in a separate one. Does this sound acceptable?

Yeah, that sounds great. Thanks a lot, @peter-toth!

> There are 2 aggregates in both subqueries so without dedup both (2) and this PR could cause regressions.

IIUC, I think it sounds like an existing bug for struct subfield pruning, which could be blocking (2) but is orthogonal to (2). For instance, if I write your example join query manually, I'd expect the struct subfield pruning to happen to the struct constructor, regardless of the existence of subqueries.

> I've never seen such transformations in SparkStrategys.

It's not uncommon in exploration Strategies such as index selection, common subplan dedup, when we substitute the subtree of a tree node T with another subtree (from somewhere else in the plan or a different access path) that may contain unneeded columns for T.

Reply:
Thanks @sigmod. I will try to update the PR by end of this week.

Reply:
Sorry, I got a bit distracted. But updated the PR to follow your (1) suggestion now. Please let me know your thoughts.